core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java

```diff
@@ -0,0 +1,375 @@
+     * @return the shards on which to execute the operation
+     */
+    protected abstract GroupShardsIterator shards(ClusterState clusterState, Request request, String[] concreteIndices);
+
+    /**
+     * Executes a global block check before polling the cluster state.
+     *
+     * @param state   the cluster state
+     * @param request the underlying request
+     * @return a non-null exception if the operation is blocked
+     */
+    protected abstract ClusterBlockException checkGlobalBlock(ClusterState state, Request request);
+
+    /**
+     * Executes a global request-level check before polling the cluster state.
+     *
+     * @param state           the cluster state
+     * @param request         the underlying request
+     * @param concreteIndices the concrete indices on which to execute the operation
+     * @return a non-null exception if the operation if blocked
+     */
+    protected abstract ClusterBlockException checkRequestBlock(ClusterState state, Request request, String[] concreteIndices);
+
+    @Override
+    protected void doExecute(Request request, ActionListener<Response> listener) {
+        new AsyncAction(request, listener).start();
+    }
+
+    protected class AsyncAction {
+        private final Request request;
+        private final ActionListener<Response> listener;
+        private final ClusterState clusterState;
+        private final DiscoveryNodes nodes;
+        private final Map<String, List<ShardRouting>> nodeIds;
+        private final AtomicReferenceArray<Object> responses;
+        private final AtomicInteger counter = new AtomicInteger();
+        private List<NoShardAvailableActionException> unavailableShardExceptions = Lists.newArrayList();
+
+        protected AsyncAction(Request request, ActionListener<Response> listener) {
+            this.request = request;
+            this.listener = listener;
+
+            clusterState = clusterService.state();
+            nodes = clusterState.nodes();
+
+            ClusterBlockException globalBlockException = checkGlobalBlock(clusterState, request);
+            if (globalBlockException != null) {
+                throw globalBlockException;
+            }
+
+            String[] concreteIndices = indexNameExpressionResolver.concreteIndices(clusterState, request);
+            ClusterBlockException requestBlockException = checkRequestBlock(clusterState, request, concreteIndices);
+            if (requestBlockException != null) {
+                throw requestBlockException;
+            }
+
+            logger.trace("resolving shards based on cluster state version [{}]", clusterState.version());
+            GroupShardsIterator shardIts = shards(clusterState, request, concreteIndices);
+            nodeIds = Maps.newHashMap();
+
+            for (ShardIterator shardIt : shardIts) {
+                ShardRouting shard = shardIt.nextOrNull();
+                if (shard != null) {
+                    String nodeId = shard.currentNodeId();
```

Comment:
when we move to ShardIterator we can have shards which are not assigned (and can't have null shards)